models/src/main/scala/coop/rchain/models/rholang/implicits.scala

```diff
@@ -502,4 +502,131 @@ object implicits {
           case _ => BitSet()
         }
     }
+
+  implicit val ChannelIsForgeable: IsForgeable[Channel] =
+    new IsForgeable[Channel] {
+      def isForgeable(chan: Channel) =
+        chan.channelInstance match {
+          case Quote(p) => ParIsForgeable(p)
+          case _        => true
+        }
+    }
+
+  implicit val ParIsForgeable: IsForgeable[Par] =
+    new IsForgeable[Par] {
+      def isForgeable(par: Par) =
+        par.ids.isEmpty &&
+          par.sends.forall(SendIsForgeable.apply) &&
+          par.receives.forall(ReceiveIsForgeable.apply) &&
+          par.news.forall(NewIsForgeable.apply) &&
+          par.exprs.forall(ExprIsForgeable.apply) &&
+          par.matches.forall(MatchIsForgeable.apply) &&
+          par.bundles.forall(BundleIsForgeable.apply) &&
+          par.connectives.forall(ConnectiveIsForgeable.apply)
```

Comment:
I see no way of telling whether all relevant properties of every AST variant are included here.

Pattern matching kind of helps with that, because one can at least see if all the (relevant) things on the left have been used on the right.

Ideally, we'd derive those typeclasses using a scheme of 'a product is fogreable if all its components are', with the sole exception of GPrivate - which isn't forgrable.

I'm not saying we should do neither of those now.

Reply:
> Ideally, we'd derive those typeclasses using a scheme of 'a product is fogreable if all its components are', with the sole exception of GPrivate - which is unforgrable.

This idea sounds cool but I am not sure whether it's doable at the moment. Maybe once you finish with your work and we have better understanding of type class derivation and overwriting protos we can do that.

Reply:
On the plus side, the polarity matters, so the default will be that things are forgeable.
So if we add a new construct, it will be obviously broken, rather than un-obviously insecure.

Reply:
I didn't do it on purpose, but I might have re-arranged it after this discussion if it went the other way.